### PR TITLE
Allow ignore to be a Function

### PR DESCRIPTION
**Pull request by paulcbetts**
This PR allows the `ignore` option to be a Function as well as a Regexp

### CODE REVIEW

File: test/basic.js [hunk 1]
```diff
@@ -346,6 +346,7 @@ util.testAllPlatforms('prune test', createPruneTest)
 util.testAllPlatforms('ignore test: string in array', createIgnoreTest, ['ignorethis'], 'ignorethis.txt')
 util.testAllPlatforms('ignore test: string', createIgnoreTest, 'ignorethis', 'ignorethis.txt')
 util.testAllPlatforms('ignore test: RegExp', createIgnoreTest, /ignorethis/, 'ignorethis.txt')
+util.testAllPlatforms('ignore test: RegExp', createIgnoreTest, function (file) { return !(file.match(/ignorethis/)) }, 'ignorethis.txt')
```
Comment (malept) [inline test/basic.js]: This should probably be named differently.

Comment (paulcbetts) [inline test/basic.js]: hwhoops, agree

**Comment by malept** [comment]
Thanks for the PR! I have no objections to this new feature, does any other contributor have any constructive criticism?

(@paulcbetts, once there's a consensus, could you please squash your commits?)

**Comment by kfranqueiro** [comment]
My one concern is that specifying `ignore` as a function, but then requiring it to return `false` to *ignore*, feels kind of backwards.  (This is primarily awkward because previously the function was internal whereas now it's directly exposed.)  I realize the doc update explains that, but I figure it's still apt to throw people.

Given that the code calls the function within the function it actually passes to `filter` anyway, could we throw `!` in front of that call so that `ignore` as a function could be implemented to return `true` to ignore instead?

**Comment by kfranqueiro** [comment]
Also, it seems as if the added test failed on CI?

**Comment by paulcbetts** [comment]
> (@paulcbetts, once there's a consensus, could you please squash your commits?)

nbd

> Given that the code calls the ignore function within the function it actually passes to filter anyway, could we throw ! in front of that call so that ignore as a function could be implemented to return true to ignore instead?

I'm okay with that

> Also, it seems as if the added test failed on CI?

Yeah, I gotta fix that up

**Comment by maxogden** [comment]
lgtm! only nit is that `name` is a reserved keyword so I normally don't use it, but it probably doesn't matter here

**Pull request merged**